# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Selenium Doc](https://www.selenium.dev/documentation/)
- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `Selenium` and `pandas` are imported for you.

In [ ]:
#pip install selenium
#%pip install webdriver-manager

In [57]:
import pandas as pd
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time
import warnings
warnings.filterwarnings('ignore')
from selenium.webdriver.common.keys import Keys
import re
import urllib.request
from selenium.webdriver.common.by import By


#### Enable the options you may need. In the next cell you have an example of them but you can choose to use them or not.

In [58]:
#driver configuration
opciones=Options()

opciones.add_experimental_option('excludeSwitches', ['enable-automation'])
opciones.add_experimental_option('useAutomationExtension', False)
opciones.headless=False    # si True, no aperece la ventana (headless=no visible)
opciones.add_argument('--start-maximized')         # comienza maximizado
#opciones.add_argument('user-data-dir=selenium')    # mantiene las cookies
#opciones.add_extension('driver_folder/adblock.crx')       # adblocker
opciones.add_argument('--incognito')

In [161]:
PATH = ChromeDriverManager().install()    # instala driver de chrome

driver = webdriver.Chrome(PATH)      # abre una ventana de chrome

driver.get('https://www.google.es')

#### Download, parse, and print the content from the Trending Developers page from GitHub:

In [60]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [61]:
#your code

driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(url)




In [62]:
DeveloperBox = driver.find_element(By.XPATH, '/html/body/div[1]/div[4]/main/div[3]/div')
DeveloperBox

<selenium.webdriver.remote.webelement.WebElement (session="f974f063ad27c585022315b5a03c5264", element="981f7445-6100-445c-a62a-499738a87d07")>

In [63]:
Developer = DeveloperBox.find_elements(By.CLASS_NAME, 'h3.lh-condensed')

[element.text for element in Developer]

['Mattias Wadman',
 'lllyasviel',
 'Lianmin Zheng',
 'Jerry Liu',
 'Henrik Rydgård',
 'Harrison Chase',
 'Fons van der Plas',
 'Steven Tey',
 'Andreas Thomas',
 'pilcrowOnPaper',
 'Laurent Mazare',
 'yetone',
 'Tom Payne',
 'Raphaël Benitte',
 'Dillon Nys',
 'Will McGugan',
 'Luca Palmieri',
 'Payton Swick',
 'Martí Climent',
 'dgtlmoon',
 'Jelle Zijlstra',
 'J. Nick Koston',
 'Nuno Campos',
 'Juliette',
 'Lee Robinson']

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [ ]:
#your code

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [64]:
# This is the url you will scrape in this exercise
urlRepo = 'https://github.com/trending/python?since=daily'

In [65]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(urlRepo)




In [66]:
RepoBox= driver.find_element(By.XPATH, '/html/body/div[1]/div[4]/main/div[3]/div')
RepoBox.text


'Repositories\nDevelopers\nSpoken Language: Any\nLanguage: Python\nDate range: Today\nmlc-ai / mlc-llm\nEnable everyone to develop, optimize and deploy AI models natively on everyone\'s devices.\nPython 3,966 225 Built by\n1,275 stars today\nAIGC-Audio / AudioGPT\nAudioGPT: Understanding and Generating Speech, Music, Sound, and Talking Head\nPython 6,496 559 Built by\n760 stars today\ngventuri / pandas-ai\nPandas AI is a Python library that integrates generative artificial intelligence capabilities into Pandas, making dataframes conversational\nPython 1,489 86 Built by\n463 stars today\nxtekky / gpt4free\ndecentralising the Ai Industry, just some language model api\'s...\nPython 25,926 5,367 Built by\n2,769 stars today\nirgolic / AutoPR\nFix issues with AI-generated pull requests, powered by ChatGPT\nPython 585 41 Built by\n71 stars today\nlamini-ai / lamini\nPython 1,275 81 Built by\n185 stars today\ndeep-floyd / IF\nPython 4,970 288 Built by\n283 stars today\ndeforum-art / deforum-st

In [67]:
Repos = RepoBox.find_elements(By.CLASS_NAME, 'h3.lh-condensed')

[element.text for element in Repos]

['mlc-ai / mlc-llm',
 'AIGC-Audio / AudioGPT',
 'gventuri / pandas-ai',
 'xtekky / gpt4free',
 'irgolic / AutoPR',
 'lamini-ai / lamini',
 'deep-floyd / IF',
 'deforum-art / deforum-stable-diffusion',
 'home-assistant / core',
 'cirediatpl / FigmaChain',
 'IntelligenzaArtificiale / Free-AUTO-GPT-with-NO-API',
 'mlc-ai / web-llm',
 'AUTOMATIC1111 / stable-diffusion-webui',
 'ddPn08 / Radiata',
 'deforum-art / deforum-for-automatic1111-webui',
 'ZrrSkywalker / LLaMA-Adapter',
 'yt-dlp / yt-dlp',
 'nlpxucan / WizardLM',
 'GitHubDaily / ChatGPT-Prompt-Engineering-for-Developers-in-Chinese',
 'georgia-tech-db / eva',
 'xtekky / chatgpt-clone',
 'mahdibland / V2RayAggregator',
 'KurtBestor / Hitomi-Downloader',
 'jaywalnut310 / vits',
 'isaiahbjork / Auto-GPT-Crypto-Plugin']

#### Display all the image links from Walt Disney wikipedia page

In [69]:
# This is the url you will scrape in this exercise
urlDis = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [70]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(urlDis)


In [116]:

DisBox= driver.find_element(By.XPATH, '//*[@id="mw-content-text"]/div[1]')
DisBox

<selenium.webdriver.remote.webelement.WebElement (session="47e6694d1e9c18ed4034e63eb7ce29d2", element="1ba26a6e-fcf4-4f3d-89d7-75ac28245169")>

In [119]:

Disney = DisBox.find_elements(By.TAG_NAME, 'img')

Disimgs= []
for element in Disney:
    src = element.get_attribute('src')
    Disimgs.append(src)

Disimgs



['https://upload.wikimedia.org/wikipedia/en/thumb/c/cb/Flyingcircus_2.jpg/250px-Flyingcircus_2.jpg',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/c/cc/The_Four_Yorkshiremen%2C_2014_%28crop%29.jpg/220px-The_Four_Yorkshiremen%2C_2014_%28crop%29.jpg',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/a/a3/Monty_Python_Live_02-07-14_12_56_41_%2814415567757%29.jpg/220px-Monty_Python_Live_02-07-14_12_56_41_%2814415567757%29.jpg',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/47/Monty_Python_Live_02-07-14_12_46_43_%2814415411808%29.jpg/220px-Monty_Python_Live_02-07-14_12_46_43_%2814415411808%29.jpg',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/a/ab/Monty_python_foot.png/220px-Monty_python_foot.png',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/2/25/Monty_Python_Live_02-07-14_12_24_16_%2814598681721%29.jpg/220px-Monty_Python_Live_02-07-14_12_24_16_%2814598681721%29.jpg',
 'https://upload.wikimedia.org/wikipedia/commons/thumb/8/89/Lumebrjack_Song_O

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [132]:
# This is the url you will scrape in this exercise
'https://en.wikipedia.org/wiki/Python' 

urlPy = 'https://en.wikipedia.org/wiki/Monty_Python'

In [133]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(urlPy)


In [134]:
PyBox= driver.find_element(By.XPATH, '//*[@id="mw-content-text"]/div[1]')
PyBox

<selenium.webdriver.remote.webelement.WebElement (session="e76ec50341d66b61e55afa98e6a73c77", element="ec79f89a-60d8-4c3e-883d-032e073c16d9")>

In [140]:
Python = PyBox.find_elements(By.TAG_NAME, 'a')

Pytimgs= []
for element in Python:
    src = element.get_attribute('href')
    Pytimgs.append(src)

Pytimgs

['https://en.wikipedia.org/wiki/Monty_Python%27s_Flying_Circus',
 'https://en.wikipedia.org/wiki/Pythonesque_(play)',
 'https://en.wikipedia.org/wiki/The_Pythons_(film)',
 'https://en.wikipedia.org/wiki/File:Flyingcircus_2.jpg',
 'https://en.wikipedia.org/wiki/Monty_Python#cite_note-1',
 'https://en.wikipedia.org/wiki/Comedic_genres',
 'https://en.wikipedia.org/wiki/Satire',
 'https://en.wikipedia.org/wiki/Surreal_humour',
 'https://en.wikipedia.org/wiki/Black_comedy',
 'https://en.wikipedia.org/wiki/Wordplay',
 'https://en.wikipedia.org/wiki/Wit',
 'https://en.wikipedia.org/wiki/Graham_Chapman',
 'https://en.wikipedia.org/wiki/John_Cleese',
 'https://en.wikipedia.org/wiki/Terry_Gilliam',
 'https://en.wikipedia.org/wiki/Eric_Idle',
 'https://en.wikipedia.org/wiki/Terry_Jones',
 'https://en.wikipedia.org/wiki/Michael_Palin',
 'http://montypython.com/',
 'https://en.wikipedia.org/wiki/Monty_Python#cite_note-2',
 'https://en.wikipedia.org/wiki/Monty_Python#cite_note-3',
 'https://en.wikip

#### Number of Titles that have changed in the United States Code since its last release point 

In [156]:
# This is the url you will scrape in this exercise
USCurl = 'http://uscode.house.gov/download/download.shtml'

In [157]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(USCurl)


In [159]:



'''USCBox= driver.find_element(By.XPATH, '//*[@id="content"]/div')
USCBox'''

USCBox = driver.find_elements(By.CLASS_NAME, 'usctitlechanged')
len(USCBox)

37

#### A Python list with the top ten FBI's Most Wanted names 

In [163]:
# This is the url you will scrape in this exercise
urlFBI = 'https://www.fbi.gov/wanted/topten'

In [165]:
#your code 
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(urlFBI)


In [172]:
FBIBox= driver.find_element(By.XPATH, '//*[@id="query-results-0f737222c5054a81a120bce207b0446a"]/ul')
FBIBox

<selenium.webdriver.remote.webelement.WebElement (session="b43c6da4aa3e952245de65b538f0371e", element="d1ffa87d-dfe2-45f2-91c0-cb24909c885e")>

In [175]:
FBI = FBIBox.find_elements(By.TAG_NAME, 'h3')

FBInames = []
for element in FBI:
    name_elements = element.find_elements(By.TAG_NAME, 'a')
    for name_element in name_elements:
        name = name_element.text
        FBInames.append(name)

FBInames

['BHADRESHKUMAR CHETANBHAI PATEL',
 'WILVER VILLEGAS-PALOMINO',
 'ALEJANDRO ROSALES CASTILLO',
 'RUJA IGNATOVA',
 'ARNOLDO JIMENEZ',
 'OMAR ALEXANDER CARDENAS',
 'ALEXIS FLORES',
 'YULAN ADONAY ARCHAGA CARIAS',
 'MICHAEL JAMES PRATT',
 'JOSE RODOLFO VILLARREAL-HERNANDEZ']

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

In [ ]:
#your code

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [179]:
# This is the url you will scrape in this exercise
urlanguage = 'https://www.wikipedia.org/'

In [187]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(urlanguage)


In [188]:


LangBox= driver.find_element(By.XPATH, '//*[@id="www-wikipedia-org"]/div[2]')
LangBox

<selenium.webdriver.remote.webelement.WebElement (session="c71b6f774d66ff10134d56f419973ce4", element="079116a3-3870-48bc-bd76-d2050322e8b6")>

In [200]:
'''Languages = LangBox.find_elements(By.CLASS_NAME, 'central-featured')

Languagenames = []
for element in Languages:
    Lang_elements = element.find_elements(By.CLASS_NAME, 'central-featured-lang')
    for Lang_element in Lang_elements:
        Lang = Lang_element.get_attribute('lang').text
        Languagenames.append(Lang)

len(Languagenames)'''

0

In [203]:
LanguageBox= driver.find_elements(By.CLASS_NAME, 'central-featured-lang')

for element in LanguageBox:
    print(element.text)

English
6 644 000+ articles
Русский
1 909 000+ статей
日本語
1 370 000+ 記事
Deutsch
2 792 000+ Artikel
Español
1 854 000+ artículos
Français
2 514 000+ articles
Italiano
1 806 000+ voci
中文
1 347 000+ 条目 / 條目
فارسی
959 000+ مقاله
Português
1 101 000+ artigos


#### A list with the different kind of datasets available in data.gov.uk 

In [205]:
# This is the url you will scrape in this exercise
UKurl = 'https://data.gov.uk/'

In [208]:
#your code 

driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(UKurl)

In [215]:
UKsets = driver.find_elements(By.CLASS_NAME, 'govuk-link')
UKsets_text = [element.text for element in UKsets]
UKsets_text


['cookies to collect information',
 'View cookies',
 '',
 'feedback',
 'Business and economy',
 'Crime and justice',
 'Defence',
 'Education',
 'Environment',
 'Government',
 'Government spending',
 'Health',
 'Mapping',
 'Society',
 'Towns and cities',
 'Transport',
 'Digital service performance',
 'Government reference data']

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [249]:
# This is the url you will scrape in this exercise
Nativeurl = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [250]:
#your code
driver=webdriver.Chrome(PATH)       # abre una venta una de chrome

driver.get(Nativeurl)

In [259]:
Nativetable = driver.find_element(By.CLASS_NAME, 'wikitable')

In [260]:
rows = Nativetable.find_elements(By.TAG_NAME, 'tr')


In [261]:
data = []
count = 0
for row in rows:
    if count >= 10:
        break
    cells = row.find_elements(By.TAG_NAME, 'td')
    if len(cells) > 0:
        link = cells[0].find_element(By.TAG_NAME, 'a')
        text = link.text
        data.append(text)
        count += 1

In [262]:
Nativetop = pd.DataFrame(data, columns=['Language'])


In [263]:
import pandas as pd

table = driver.find_element(By.CLASS_NAME, 'wikitable')
rows = table.find_elements(By.TAG_NAME, 'tr')

data = []
count = 0
for row in rows:
    if count >= 10:
        break
    cells = row.find_elements(By.TAG_NAME, 'td')
    if len(cells) > 0:
        link = cells[0].find_element(By.TAG_NAME, 'a')
        text = link.text
        data.append(text)
        count += 1

Nativetop = pd.DataFrame(data, columns=['Language'])

df

Language
0  Mandarin Chinese
1           Spanish
2           English
3             Hindi
4        Portuguese
5           Bengali
6           Russian
7          Japanese
8       Yue Chinese
9        Vietnamese

In [228]:
NativeBox= driver.find_elements(By.CLASS_NAME, 'wikitable sortable static-row-numbers- jquery-tablesorter')
NativeBox

[]

In [229]:
NativeLang = []
for element in NativeBox:
    Langs = element.find_elements(By.TAG_NAME, 'href')
    for Lang in Langs:
        name = Lang.text
        NativeLang.append(name)

NativeLang

[]

In [ ]:


FBInames = []
for element in FBI:
    name_elements = element.find_elements(By.TAG_NAME, 'a')
    for name_element in name_elements:
        name = name_element.text
        FBInames.append(name)


### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code